raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1914,8 +1917,7 @@ private long pollLeader(long currentTimeMs) {
         LeaderState<T> state = quorum.leaderStateOrThrow();
         maybeFireLeaderChange(state);
 
-        GracefulShutdown shutdown = this.shutdown.get();
-        if (shutdown != null) {
+        if (shutdown.get() != null || resignedEpoch.get() == state.epoch()) {
```

Comment:
It seems that we handle `resign` and `shutdown` without distinction, what's their difference?

Reply:
The main difference is the behavior of the Resigned state itself. When the client is shutting down, we will not transition out of the resigned state (or any other state) to become a candidate.